tools/PI/DevHome.PI/Controls/AddToolControl.xaml.cs

```diff
@@ -101,55 +161,290 @@ private void ClearValues()
+        }
+
+        if (shortcuts is not null)
+        {
+            count = shortcuts.Count;
+        }
+
+        return count;
     }
 
-    private void UpdateSampleCommandline()
+    private int GetPackages()
     {
-        if (SampleCommandTextBox is null)
+        var count = 0;
+        var packageManager = new PackageManager();
+        packages = packageManager.FindPackagesForUserWithPackageTypes(string.Empty, PackageTypes.Main).ToList();
+        if (packages is not null)
+        {
+            count = packages.Count;
+        }
+
+        return count;
+    }
+
+    private async Task ProcessItemsAsync(IProgress<int> progress)
+    {
+        // Recursively search for .lnk files.
+        var currentCount = 0;
+        for (var i = 0; i < shortcuts?.Count; i++)
+        {
+            await Task.Run(() => ProcessShortcut(shortcuts[i]));
+
+            // Report progress.
+            currentCount++;
+            var percentComplete = (i + 1) * ShortcutProcessingEndIndex / itemCount;
+            progress.Report(percentComplete);
+        }
+
+        for (var j = 0; j < packages?.Count; j++)
+        {
+            await Task.Run(() => ProcessPackage(packages[j]));
+
+            // Report progress.
+            currentCount++;
+            var percentComplete = ShortcutProcessingEndIndex + ((j + 1) * ShortcutProcessingEndIndex / itemCount);
+            progress.Report(percentComplete);
+        }
+
+        allApps = allApps.OrderBy(app => app.Name).ToList();
+
+        // We get the icon data on the UI thread, because BitmapImages must be created on the UI thread.
+        for (var k = 0; k < allApps.Count; k++)
+        {
+            var app = allApps[k];
+            if (app.IsMsix)
+            {
+                if (app.AppPackage is not null)
+                {
+                    try
+                    {
+                        // The package might be in a bad state, and accessing its
+                        // properties might throw an exception.
+                        app.Icon = new BitmapImage(app.AppPackage.Logo);
+                        app.IconFilePath = app.AppPackage.Logo.LocalPath;
+                    }
+                    catch
+                    {
+                    }
+                }
+            }
+            else
+            {
+                if (!string.IsNullOrEmpty(app.TargetPath))
+                {
+                    app.Icon = WindowHelper.GetBitmapImageFromFile(app.TargetPath);
+                }
+            }
+
+            currentCount++;
+            var percentComplete = PackageProcessingEndIndex + ((k + 1) * ShortcutProcessingEndIndex / itemCount);
+            progress.Report(percentComplete);
+
+            // Yield to make sure the UI thread can update the progress output.
+            await Task.Delay(1);
+        }
+    }
+
+    public void ProcessShortcut(string filePath)
```

Comment:
Prefer either returning a new instance of `AppInstallInfo` so the caller can decide what to do with it, or add a comment mentioning that allApps is modified.

Reply:
Added comments to both ProcessShortcut and ProcessPackage.